modular_zubbers/code/_onclick/hud/screen_objects/hud_timer.dm

```diff
@@ -0,0 +1,213 @@
+/atom/movable/screen/text
+	name = null
+	icon = null
+	icon_state = null
+	mouse_opacity = MOUSE_OPACITY_TRANSPARENT
+	screen_loc = "CENTER-7,CENTER-7"
+	maptext_height = 480
+	maptext_width = 480
+
+/// A screen object that shows the time left on a timer
+/atom/movable/screen/text/screen_timer
+	screen_loc = "CENTER-7,CENTER-7"
+	/// Left side of the HTML tag for maptext, style is also included
+	var/maptext_style_left = "<span class='maptext' style=font-size:6pt;text-align:center; align='top'>"
+	/// End tag of the HTML tag for maptext
+	var/maptext_style_right = "</span>"
+	/// The actual displayed content of the maptext, use ${timer}, and it'll be replaced with formatted time left
+	var/maptext_string
+	/// Timer ID that we're tracking, the time left of this is displayed as maptext
+	var/timer_id
+	/// The list of mobs in whose client.screens we are added to
+	var/list/timer_mobs = list()
+
+/atom/movable/screen/text/screen_timer/Initialize(
+		mapload,
+		list/mobs,
+		timer,
+		text,
+		offset_x = 150,
+		offset_y = -70,
+		style_start,
+		style_end,
+	)
+	. = ..(mapload, null)
+
+	if(!islist(mobs) && mobs)
+		mobs = list(mobs)
+	// Copy the list just in case the arguments list is a list we don't want to modify
+	if(length(mobs))
+		mobs = mobs.Copy()
+	if(!timer)
+		return INITIALIZE_HINT_QDEL
+	if(style_start)
+		maptext_style_left = style_start
+	if(style_end)
+		maptext_style_right = style_end
+	maptext_string = text
+	timer_id = timer
+	maptext_x = offset_x
+	maptext_y = offset_y
+	update_maptext()
+	if(length(mobs))
+		apply_to(mobs)
+
+/atom/movable/screen/text/screen_timer/process()
+	if(!timeleft(timer_id))
+		delete_self()
+		return
+	update_maptext()
+
+/// Adds the object to the client.screen of all mobs in the list, and registers the needed signals
+/// TODO: WEAKREFS ARE FUCKY WITH REMOVING
+/atom/movable/screen/text/screen_timer/proc/apply_to(list/mobs)
+	if(!islist(mobs))
+		if(!mobs)
+			return
+		mobs = list(mobs)
+	if(!length(timer_mobs) && length(mobs))
+		START_PROCESSING(SSprocessing, src)
+	for(var/player in mobs)
+		if(player in timer_mobs)
+			continue
+		if(istype(player, /datum/weakref))
+			var/datum/weakref/ref = player
+			player = ref.resolve()
+		attach(player)
+		RegisterSignal(player, COMSIG_MOB_LOGIN, PROC_REF(attach))
+		timer_mobs += WEAKREF(player)
+
+/// Removes the object from the client.screen of all mobs in the list, and unregisters the needed signals, while also stopping processing if there's no more mobs in the screen timers mob list
+/atom/movable/screen/text/screen_timer/proc/remove_from(list/mobs)
+	if(!islist(mobs))
+		if(!mobs)
+			return
+		mobs = list(mobs)
+	for(var/player in mobs)
+		UnregisterSignal(player, COMSIG_MOB_LOGIN)
+		if(player in timer_mobs)
+			timer_mobs -= player
+		if(istype(player, /datum/weakref))
+			var/datum/weakref/ref = player
+			player = ref.resolve()
+		de_attach(player)
+	if(!length(timer_mobs))
+		STOP_PROCESSING(SSprocessing, src)
```

Comment:
There is a bit of bad code here.

``if(!mobs) return`` should be placed imediately after the proc declaration

``if(player in timer_mobs)`` is not needed, you can just call ``timer_mobs -= player`` and it will do it for you.

``if(istype(player, /datum/weakref))`` is also redundent, as the "mobs" list should always contain weakrefs

Reply:
The intent was to allow to remove things without a weakref, but that won't work as that's a entirely different object if you simply make a new weakref on the same target, might work if I resolve and them compare them

Reply:
reworked remove_from heavily